controller/api/destination/traffic_split_adaptor.go

```diff
@@ -0,0 +1,69 @@
+package destination
+
+import (
+	"fmt"
+
+	ts "github.com/deislabs/smi-sdk-go/pkg/apis/split/v1alpha1"
+	"github.com/linkerd/linkerd2/controller/api/destination/watcher"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+)
+
+// trafficSplitAdaptor merges traffic splits into service profiles, encoding
+// them as dst overrides.  trafficSplitAdaptor holds an underlying
+// ProfileUpdateListener and updates that listener with a merged service
+// service profile which includes the traffic split logic as a dst override
+// when a traffic split exists.  trafficSplitAdaptor itself implements both
+// ProfileUpdateListener and TrafficSplitUpdateListener and must be passed to
+// a source of profile updates (such as a ProfileWatcher) and a source of
+// traffic split updates (such as a TrafficSplitWatcher).
+type trafficSplitAdaptor struct {
+	listener watcher.ProfileUpdateListener
+	id       watcher.ServiceID
+	port     watcher.Port
+	profile  *sp.ServiceProfile
+	split    *ts.TrafficSplit
+}
+
+func newTrafficSplitAdaptor(listener watcher.ProfileUpdateListener, id watcher.ServiceID, port watcher.Port) *trafficSplitAdaptor {
+	return &trafficSplitAdaptor{
+		listener: listener,
+		id:       id,
+		port:     port,
+	}
+}
+
+func (tsa *trafficSplitAdaptor) Update(profile *sp.ServiceProfile) {
+	tsa.profile = profile
+	tsa.publish()
+}
+
+func (tsa *trafficSplitAdaptor) UpdateTrafficSplit(split *ts.TrafficSplit) {
+	if tsa.split != split {
```

Comment:
This is a pointer comparison, which I'm guessing might not be what you want. Maybe use `reflect.DeepEqual`?

Reply:
This was a quick and dirty check to avoid sending traffic split updates when the traffic split is updated from `nil` to `nil`.  This helps to avoid duplicate updates in most cases (although there are still duplicate updates in some cases, as you described).  I'll tweak this to be more explicit about what it's doing.